File: Problems/621/621_Task_Scheduler.py
from typing import List
from collections import Counter
from heapq import heappush, heappop


class Solution:
    def leastInterval(self, tasks: List[str], n: int) -> int:
        table = [0]*26
        for c in tasks:
            table[ord(c) - ord('A')] += 1

        maxCnt = max(table)
        maxCntTaskCnt = 0
        for cnt in table:
            if cnt == maxCnt:
                maxCntTaskCnt += 1
        return max((n+1) * (maxCnt-1) + maxCntTaskCnt, len(tasks))


sol = Solution()
tasks = ["A","A","A","B","B","B", "C","C","C", "D", "D", "E"]
print(sol.leastInterval(tasks, 2))

File: Problems/621/621_Task_Scheduler.cpp
class Solution {
public:
    int leastInterval(vector<char>& tasks, int n) {
        vector<int>taskCnt(26, 0);
        for(char c: tasks){
            taskCnt[c-'A']++;
        }
        int maxCnt = 0;
        int maxCntTaskCnt = 0;
        for(int cnt: taskCnt){
            if(cnt>maxCnt){
                maxCnt = cnt;
                maxCntTaskCnt = 1;
            }else if(cnt==maxCnt){
                maxCntTaskCnt++;
            }
        }
        int ans = (maxCnt-1)*(n+1) + maxCntTaskCnt;
        return ans>tasks.size()?ans:tasks.size();
    }
};